{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "588f0c2d-3b32-4865-a7b6-b5db5776ada4",
   "metadata": {},
   "source": [
    "## Sanyukta Chapagain\n",
    "## Machine Learning in Bioinformatics\n",
    "## Homework 2\n",
    "## Programming 2 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45d3b57b-4f9f-4905-913a-130bd8d53e7c",
   "metadata": {},
   "source": [
    "#### In this project, we worked with the BRCA_TCGA_PUB2015 dataset from The Cancer Genome Atlas (TCGA), specifically focused on breast cancer patients.This dataset was obtained as a compressed file: brca_tcga_pub2015.tar.gz After extraction, it provided multiple data types including:\n",
    "Mutation data (data_mutations.txt)\n",
    "\n",
    "Clinical data (data_clinical_sample.txt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d30cb11e-ca19-4ff2-b634-25d36274eba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import zipfile\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d97fcfc6-bca7-4365-83d7-7694491f19c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Extracted .tar.gz successfully!\n"
     ]
    }
   ],
   "source": [
    "import tarfile\n",
    "import os\n",
    "\n",
    "# Path to the .tar.gz file\n",
    "tarPath = \"brca_tcga_pub2015.tar.gz\"\n",
    "\n",
    "# Open and extract\n",
    "if tarfile.is_tarfile(tarPath):\n",
    "    with tarfile.open(tarPath, \"r:gz\") as tar:\n",
    "        tar.extractall(\"brca_tcga_pub2015\")\n",
    "        print(\" Extracted .tar.gz successfully!\")\n",
    "else:\n",
    "    print(\" Not a valid tar.gz file.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8dcc6c23-1032-4844-a0b5-c00fe727a202",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['case_lists',\n",
       " 'data_clinical_patient.txt',\n",
       " 'data_clinical_sample.txt',\n",
       " 'data_cna.txt',\n",
       " 'data_cna_hg19.seg',\n",
       " 'data_gistic_genes_amp.txt',\n",
       " 'data_gistic_genes_del.txt',\n",
       " 'data_linear_cna.txt',\n",
       " 'data_methylation_hm27.txt',\n",
       " 'data_methylation_hm450.txt',\n",
       " 'data_mrna_agilent_microarray.txt',\n",
       " 'data_mrna_agilent_microarray_zscores_ref_all_samples.txt',\n",
       " 'data_mrna_agilent_microarray_zscores_ref_diploid_samples.txt',\n",
       " 'data_mrna_seq_v2_rsem.txt',\n",
       " 'data_mrna_seq_v2_rsem_zscores_ref_all_samples.txt',\n",
       " 'data_mrna_seq_v2_rsem_zscores_ref_diploid_samples.txt',\n",
       " 'data_mutations.txt',\n",
       " 'data_mutsig.txt',\n",
       " 'data_rppa.txt',\n",
       " 'data_rppa_zscores.txt',\n",
       " 'LICENSE',\n",
       " 'meta_clinical_patient.txt',\n",
       " 'meta_clinical_sample.txt',\n",
       " 'meta_cna.txt',\n",
       " 'meta_cna_hg19_seg.txt',\n",
       " 'meta_linear_cna.txt',\n",
       " 'meta_methylation_hm27.txt',\n",
       " 'meta_methylation_hm450.txt',\n",
       " 'meta_mrna_agilent_microarray.txt',\n",
       " 'meta_mrna_agilent_microarray_zscores_ref_all_samples.txt',\n",
       " 'meta_mrna_agilent_microarray_zscores_ref_diploid_samples.txt',\n",
       " 'meta_mrna_seq_v2_rsem.txt',\n",
       " 'meta_mrna_seq_v2_rsem_zscores_ref_all_samples.txt',\n",
       " 'meta_mrna_seq_v2_rsem_zscores_ref_diploid_samples.txt',\n",
       " 'meta_mutations.txt',\n",
       " 'meta_rppa.txt',\n",
       " 'meta_rppa_zscores.txt',\n",
       " 'meta_study.txt']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check what's inside the extracted folder\n",
    "unzippedPath = \"brca_tcga_pub2015/brca_tcga_pub2015/\"\n",
    "os.listdir(unzippedPath)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b02bbcc-632c-4918-bcbf-2b0b5debd8e8",
   "metadata": {},
   "source": [
    "## File preprocessing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ce049ccb-0af2-484b-b05d-5b7964289f2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import libraries\n",
    "\n",
    "#Basic data handling libraries\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt, seaborn as sns\n",
    "%matplotlib inline\n",
    "sns.set(rc={'axes.facecolor':'lightblue'})\n",
    "\n",
    "#Scikitlearn libraries\n",
    "import sklearn\n",
    "from sklearn import metrics\n",
    "from sklearn import model_selection\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import classification_report, precision_recall_curve\n",
    "from sklearn.model_selection import StratifiedKFold, train_test_split, RepeatedStratifiedKFold, GridSearchCV, cross_val_score\n",
    "from sklearn.linear_model import LinearRegression, LogisticRegression\n",
    "from sklearn.feature_selection import RFE\n",
    "from sklearn.impute import KNNImputer\n",
    "from sklearn.datasets import make_blobs\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import make_scorer, accuracy_score, precision_score, recall_score, f1_score\n",
    "from sklearn.naive_bayes import GaussianNB, BernoulliNB\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n",
    "\n",
    "\n",
    "# Statistics import\n",
    "from statistics import mean, stdev\n",
    "\n",
    "\n",
    "# Importing statsmodels\n",
    "import statsmodels.api as sm\n",
    "\n",
    "# Importing 'variance_inflation_factor'\n",
    "from statsmodels.stats.outliers_influence import variance_inflation_factor\n",
    "\n",
    "# Importing warnings\n",
    "# ------------------\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d9302b1-3374-47d4-9b77-9d1511e99dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataPath = \"brca_tcga_pub2015/brca_tcga_pub2015/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "80e25f6c-9387-4f1c-8a56-a682743ad71a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load clinical data\n",
    "df_clinicalsample = pd.read_csv(dataPath + \"data_clinical_sample.txt\", sep=\"\\t\", skiprows=4)\n",
    "df_clinicalsample.replace('[Not Available]', np.nan, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9ca76a90-ef79-46f8-8d87-625976abc8e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th>OTHER_SAMPLE_ID</th>\n",
       "      <th>DAYS_TO_COLLECTION</th>\n",
       "      <th>IS_FFPE</th>\n",
       "      <th>OCT_EMBEDDED</th>\n",
       "      <th>PATHOLOGY_REPORT_FILE_NAME</th>\n",
       "      <th>SURGICAL_PROCEDURE_FIRST</th>\n",
       "      <th>FIRST_SURGICAL_PROCEDURE_OTHER</th>\n",
       "      <th>SURGERY_FOR_POSITIVE_MARGINS</th>\n",
       "      <th>...</th>\n",
       "      <th>METASTATIC_TUMOR_INDICATOR</th>\n",
       "      <th>PROJECT_CODE</th>\n",
       "      <th>TISSUE_SOURCE_SITE</th>\n",
       "      <th>TUMOR_TISSUE_SITE</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SOMATIC_STATUS</th>\n",
       "      <th>TMB_NONSYNONYMOUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-LQ-A4E4</td>\n",
       "      <td>TCGA-LQ-A4E4-01</td>\n",
       "      <td>8C8D4BD4-3AA9-4AD1-9715-2376EE540A0C</td>\n",
       "      <td>414</td>\n",
       "      <td>NO</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TCGA-LQ-A4E4.C13D16C9-BAC7-4CA4-B062-D925466F9...</td>\n",
       "      <td>Modified Radical Mastectomy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LQ</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "      <td>ILC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>2.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-A2-A3KC</td>\n",
       "      <td>TCGA-A2-A3KC-01</td>\n",
       "      <td>8318A9A4-E78E-4DB8-97C7-2CC733D0C512</td>\n",
       "      <td>299</td>\n",
       "      <td>NO</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TCGA-A2-A3KC.593AF241-8F84-4BA0-8878-7C1CE72A4...</td>\n",
       "      <td>Simple Mastectomy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "      <td>ILC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>0.766667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-A2-A3KD</td>\n",
       "      <td>TCGA-A2-A3KD-01</td>\n",
       "      <td>401D075C-D443-4592-8C03-E675CAAD2B50</td>\n",
       "      <td>700</td>\n",
       "      <td>NO</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TCGA-A2-A3KD.3E4717BA-E9AE-45F9-9696-F8928D9FC...</td>\n",
       "      <td>Lumpectomy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mastectomy NOS</td>\n",
       "      <td>...</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>0.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A7-A0D9</td>\n",
       "      <td>TCGA-A7-A0D9-01</td>\n",
       "      <td>c144ae50-ed29-4e27-bbee-fa81e79ac7db</td>\n",
       "      <td>173</td>\n",
       "      <td>NO</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>TCGA-A7-A0D9.7D5763F7-7284-4AF5-BBFC-5D8B5CB7F...</td>\n",
       "      <td>Simple Mastectomy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A7</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>0.900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A7-A0DA</td>\n",
       "      <td>TCGA-A7-A0DA-01</td>\n",
       "      <td>4f441e61-6bea-4a12-841d-def270804bbe</td>\n",
       "      <td>177</td>\n",
       "      <td>NO</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>TCGA-A7-A0DA.69AC5937-3FFD-40FB-9922-79DB3CED7...</td>\n",
       "      <td>Lumpectomy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A7</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>4.233333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     PATIENT_ID        SAMPLE_ID                       OTHER_SAMPLE_ID  \\\n",
       "0  TCGA-LQ-A4E4  TCGA-LQ-A4E4-01  8C8D4BD4-3AA9-4AD1-9715-2376EE540A0C   \n",
       "1  TCGA-A2-A3KC  TCGA-A2-A3KC-01  8318A9A4-E78E-4DB8-97C7-2CC733D0C512   \n",
       "2  TCGA-A2-A3KD  TCGA-A2-A3KD-01  401D075C-D443-4592-8C03-E675CAAD2B50   \n",
       "3  TCGA-A7-A0D9  TCGA-A7-A0D9-01  c144ae50-ed29-4e27-bbee-fa81e79ac7db   \n",
       "4  TCGA-A7-A0DA  TCGA-A7-A0DA-01  4f441e61-6bea-4a12-841d-def270804bbe   \n",
       "\n",
       "  DAYS_TO_COLLECTION IS_FFPE OCT_EMBEDDED  \\\n",
       "0                414      NO         TRUE   \n",
       "1                299      NO         TRUE   \n",
       "2                700      NO         TRUE   \n",
       "3                173      NO        FALSE   \n",
       "4                177      NO        FALSE   \n",
       "\n",
       "                          PATHOLOGY_REPORT_FILE_NAME  \\\n",
       "0  TCGA-LQ-A4E4.C13D16C9-BAC7-4CA4-B062-D925466F9...   \n",
       "1  TCGA-A2-A3KC.593AF241-8F84-4BA0-8878-7C1CE72A4...   \n",
       "2  TCGA-A2-A3KD.3E4717BA-E9AE-45F9-9696-F8928D9FC...   \n",
       "3  TCGA-A7-A0D9.7D5763F7-7284-4AF5-BBFC-5D8B5CB7F...   \n",
       "4  TCGA-A7-A0DA.69AC5937-3FFD-40FB-9922-79DB3CED7...   \n",
       "\n",
       "      SURGICAL_PROCEDURE_FIRST FIRST_SURGICAL_PROCEDURE_OTHER  \\\n",
       "0  Modified Radical Mastectomy                            NaN   \n",
       "1            Simple Mastectomy                            NaN   \n",
       "2                   Lumpectomy                            NaN   \n",
       "3            Simple Mastectomy                            NaN   \n",
       "4                   Lumpectomy                            NaN   \n",
       "\n",
       "  SURGERY_FOR_POSITIVE_MARGINS  ... METASTATIC_TUMOR_INDICATOR PROJECT_CODE  \\\n",
       "0                          NaN  ...                        NaN          NaN   \n",
       "1                          NaN  ...                         NO          NaN   \n",
       "2               Mastectomy NOS  ...                         NO          NaN   \n",
       "3                          NaN  ...                        NaN          NaN   \n",
       "4                          NaN  ...                        NaN          NaN   \n",
       "\n",
       "  TISSUE_SOURCE_SITE TUMOR_TISSUE_SITE                CANCER_TYPE  \\\n",
       "0                 LQ            Breast  Invasive Breast Carcinoma   \n",
       "1                 A2            Breast  Invasive Breast Carcinoma   \n",
       "2                 A2            Breast  Invasive Breast Carcinoma   \n",
       "3                 A7            Breast  Invasive Breast Carcinoma   \n",
       "4                 A7            Breast  Invasive Breast Carcinoma   \n",
       "\n",
       "                CANCER_TYPE_DETAILED ONCOTREE_CODE SAMPLE_TYPE SOMATIC_STATUS  \\\n",
       "0  Breast Invasive Lobular Carcinoma           ILC     Primary        Matched   \n",
       "1  Breast Invasive Lobular Carcinoma           ILC     Primary        Matched   \n",
       "2          Invasive Breast Carcinoma          BRCA     Primary        Matched   \n",
       "3   Breast Invasive Ductal Carcinoma           IDC     Primary        Matched   \n",
       "4   Breast Invasive Ductal Carcinoma           IDC     Primary        Matched   \n",
       "\n",
       "  TMB_NONSYNONYMOUS  \n",
       "0          2.300000  \n",
       "1          0.766667  \n",
       "2          0.266667  \n",
       "3          0.900000  \n",
       "4          4.233333  \n",
       "\n",
       "[5 rows x 63 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinicalsample.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b18d1869-6e79-4bf0-901e-b90870342134",
   "metadata": {},
   "source": [
    "##### The clinical dataset contains detailed information about each patient's tumor sample, including patient ID, sample type, surgery details, cancer subtype, and tumor mutational burden (TMB). This metadata is essential for linking clinical outcomes with mutation data and serves as the basis for our classification labels: Ductal vs. Lobular breast cancer."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7159e131-6001-4ce8-9815-6fd3623a08ae",
   "metadata": {},
   "source": [
    "## Load mutation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe43a98e-e7d9-4cd6-93d2-0a0405822ce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load mutation data\n",
    "df_mutations = pd.read_csv(dataPath + \"data_mutations.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ee364b3a-21ed-42d2-9935-911d8c2c5114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>DOMAINS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>ENSP</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>CCDS</th>\n",
       "      <th>EA_MAF</th>\n",
       "      <th>Allele</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>PUBMED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTGER3</td>\n",
       "      <td>5733</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>71512366</td>\n",
       "      <td>71512366</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant,splice_region_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Transmembrane_helices:Tmhmm,Pfam_domain:PF0000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>probably_damaging(0.997)</td>\n",
       "      <td>ENSP00000349003</td>\n",
       "      <td>L/V</td>\n",
       "      <td>CCDS655.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>1106/1943</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLG</td>\n",
       "      <td>2312</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>152285981</td>\n",
       "      <td>152285981</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Low_complexity_(Seg):Seg,PROSITE_profiles:PS50324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>probably_damaging(0.988)</td>\n",
       "      <td>ENSP00000357789</td>\n",
       "      <td>R/W</td>\n",
       "      <td>CCDS30860.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>1417/12747</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GPR52</td>\n",
       "      <td>9293</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>174417411</td>\n",
       "      <td>174417411</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>...</td>\n",
       "      <td>Transmembrane_helices:Tmhmm,Prints_domain:PR00...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000356658</td>\n",
       "      <td>I</td>\n",
       "      <td>CCDS30941.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>200/1472</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLC35F3</td>\n",
       "      <td>148641</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>234452419</td>\n",
       "      <td>234452419</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>...</td>\n",
       "      <td>Pfam_domain:PF06027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000355577</td>\n",
       "      <td>S</td>\n",
       "      <td>CCDS1600.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>1045/2891</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR2T3</td>\n",
       "      <td>343173</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>248636826</td>\n",
       "      <td>248636826</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>Pfam_domain:PF00001,Pfam_domain:PF10320,PROSIT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>benign(0.001)</td>\n",
       "      <td>ENSP00000352604</td>\n",
       "      <td>R/C</td>\n",
       "      <td>CCDS31117.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>200/1008</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id                    Center NCBI_Build Chromosome  \\\n",
       "0      PTGER3            5733  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "1         FLG            2312  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "2       GPR52            9293          genome.wustl.edu     GRCh37          1   \n",
       "3     SLC35F3          148641  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "4       OR2T3          343173          genome.wustl.edu     GRCh37          1   \n",
       "\n",
       "   Start_Position  End_Position Strand  \\\n",
       "0        71512366      71512366      +   \n",
       "1       152285981     152285981      +   \n",
       "2       174417411     174417411      +   \n",
       "3       234452419     234452419      +   \n",
       "4       248636826     248636826      +   \n",
       "\n",
       "                              Consequence Variant_Classification  ...  \\\n",
       "0  missense_variant,splice_region_variant      Missense_Mutation  ...   \n",
       "1                        missense_variant      Missense_Mutation  ...   \n",
       "2                      synonymous_variant                 Silent  ...   \n",
       "3                      synonymous_variant                 Silent  ...   \n",
       "4                        missense_variant      Missense_Mutation  ...   \n",
       "\n",
       "                                             DOMAINS MOTIF_SCORE_CHANGE  \\\n",
       "0  Transmembrane_helices:Tmhmm,Pfam_domain:PF0000...                NaN   \n",
       "1  Low_complexity_(Seg):Seg,PROSITE_profiles:PS50324                NaN   \n",
       "2  Transmembrane_helices:Tmhmm,Prints_domain:PR00...                NaN   \n",
       "3                                Pfam_domain:PF06027                NaN   \n",
       "4  Pfam_domain:PF00001,Pfam_domain:PF10320,PROSIT...                NaN   \n",
       "\n",
       "                   PolyPhen             ENSP  Amino_acids         CCDS EA_MAF  \\\n",
       "0  probably_damaging(0.997)  ENSP00000349003          L/V    CCDS655.1    NaN   \n",
       "1  probably_damaging(0.988)  ENSP00000357789          R/W  CCDS30860.1    NaN   \n",
       "2                       NaN  ENSP00000356658            I  CCDS30941.1    NaN   \n",
       "3                       NaN  ENSP00000355577            S   CCDS1600.1    NaN   \n",
       "4             benign(0.001)  ENSP00000352604          R/C  CCDS31117.1    NaN   \n",
       "\n",
       "  Allele  cDNA_position  PUBMED  \n",
       "0      C      1106/1943     NaN  \n",
       "1      A     1417/12747     NaN  \n",
       "2      A       200/1472     NaN  \n",
       "3      T      1045/2891     NaN  \n",
       "4      T       200/1008     NaN  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Preview\n",
    "df_mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d480f6b-e087-4363-abd0-89bbee6ab3e3",
   "metadata": {},
   "source": [
    "#####  The mutation dataset contains detailed information about somatic mutations found in each tumor sample. Each row represents a specific mutation in a gene, including columns for gene name (Hugo_Symbol), mutation type (Variant_Classification), chromosome position, amino acid change, predicted effect (e.g., PolyPhen score), and more. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b094bc7c-3ec6-46b3-889b-d4d177ea06a2",
   "metadata": {},
   "source": [
    "## Extracting Patient ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4d647d09-28c5-4120-8e15-45bf3fa1ecc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>ENSP</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>CCDS</th>\n",
       "      <th>EA_MAF</th>\n",
       "      <th>Allele</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>PUBMED</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTGER3</td>\n",
       "      <td>5733</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>71512366</td>\n",
       "      <td>71512366</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant,splice_region_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>probably_damaging(0.997)</td>\n",
       "      <td>ENSP00000349003</td>\n",
       "      <td>L/V</td>\n",
       "      <td>CCDS655.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>1106/1943</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLG</td>\n",
       "      <td>2312</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>152285981</td>\n",
       "      <td>152285981</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>probably_damaging(0.988)</td>\n",
       "      <td>ENSP00000357789</td>\n",
       "      <td>R/W</td>\n",
       "      <td>CCDS30860.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>1417/12747</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GPR52</td>\n",
       "      <td>9293</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>174417411</td>\n",
       "      <td>174417411</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000356658</td>\n",
       "      <td>I</td>\n",
       "      <td>CCDS30941.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>200/1472</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLC35F3</td>\n",
       "      <td>148641</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>234452419</td>\n",
       "      <td>234452419</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000355577</td>\n",
       "      <td>S</td>\n",
       "      <td>CCDS1600.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>1045/2891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR2T3</td>\n",
       "      <td>343173</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>248636826</td>\n",
       "      <td>248636826</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>benign(0.001)</td>\n",
       "      <td>ENSP00000352604</td>\n",
       "      <td>R/C</td>\n",
       "      <td>CCDS31117.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>200/1008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id                    Center NCBI_Build Chromosome  \\\n",
       "0      PTGER3            5733  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "1         FLG            2312  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "2       GPR52            9293          genome.wustl.edu     GRCh37          1   \n",
       "3     SLC35F3          148641  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "4       OR2T3          343173          genome.wustl.edu     GRCh37          1   \n",
       "\n",
       "   Start_Position  End_Position Strand  \\\n",
       "0        71512366      71512366      +   \n",
       "1       152285981     152285981      +   \n",
       "2       174417411     174417411      +   \n",
       "3       234452419     234452419      +   \n",
       "4       248636826     248636826      +   \n",
       "\n",
       "                              Consequence Variant_Classification  ...  \\\n",
       "0  missense_variant,splice_region_variant      Missense_Mutation  ...   \n",
       "1                        missense_variant      Missense_Mutation  ...   \n",
       "2                      synonymous_variant                 Silent  ...   \n",
       "3                      synonymous_variant                 Silent  ...   \n",
       "4                        missense_variant      Missense_Mutation  ...   \n",
       "\n",
       "  MOTIF_SCORE_CHANGE                  PolyPhen             ENSP Amino_acids  \\\n",
       "0                NaN  probably_damaging(0.997)  ENSP00000349003         L/V   \n",
       "1                NaN  probably_damaging(0.988)  ENSP00000357789         R/W   \n",
       "2                NaN                       NaN  ENSP00000356658           I   \n",
       "3                NaN                       NaN  ENSP00000355577           S   \n",
       "4                NaN             benign(0.001)  ENSP00000352604         R/C   \n",
       "\n",
       "          CCDS  EA_MAF Allele cDNA_position  PUBMED    PATIENT_ID  \n",
       "0    CCDS655.1     NaN      C     1106/1943     NaN  TCGA-B6-A0IG  \n",
       "1  CCDS30860.1     NaN      A    1417/12747     NaN  TCGA-B6-A0IG  \n",
       "2  CCDS30941.1     NaN      A      200/1472     NaN  TCGA-B6-A0IG  \n",
       "3   CCDS1600.1     NaN      T     1045/2891     NaN  TCGA-B6-A0IG  \n",
       "4  CCDS31117.1     NaN      T      200/1008     NaN  TCGA-B6-A0IG  \n",
       "\n",
       "[5 rows x 80 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create PATIENT_ID by removing the last 3 characters (e.g., '-01')\n",
    "df_mutations['PATIENT_ID'] = df_mutations['Tumor_Sample_Barcode'].apply(lambda x: x[:-3])\n",
    "df_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0eff5208-de73-4c32-86b4-8846fa9f41fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>...</th>\n",
       "      <th>METASTATIC_TUMOR_INDICATOR</th>\n",
       "      <th>PROJECT_CODE</th>\n",
       "      <th>TISSUE_SOURCE_SITE</th>\n",
       "      <th>TUMOR_TISSUE_SITE</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SOMATIC_STATUS</th>\n",
       "      <th>TMB_NONSYNONYMOUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTGER3</td>\n",
       "      <td>5733</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>71512366</td>\n",
       "      <td>71512366</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant,splice_region_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>1.366667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLG</td>\n",
       "      <td>2312</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>152285981</td>\n",
       "      <td>152285981</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>1.366667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GPR52</td>\n",
       "      <td>9293</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>174417411</td>\n",
       "      <td>174417411</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>1.366667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLC35F3</td>\n",
       "      <td>148641</td>\n",
       "      <td>genome.wustl.edu;unc.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>234452419</td>\n",
       "      <td>234452419</td>\n",
       "      <td>+</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>Silent</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>1.366667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR2T3</td>\n",
       "      <td>343173</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>248636826</td>\n",
       "      <td>248636826</td>\n",
       "      <td>+</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6</td>\n",
       "      <td>Breast</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Matched</td>\n",
       "      <td>1.366667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id                    Center NCBI_Build Chromosome  \\\n",
       "0      PTGER3            5733  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "1         FLG            2312  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "2       GPR52            9293          genome.wustl.edu     GRCh37          1   \n",
       "3     SLC35F3          148641  genome.wustl.edu;unc.edu     GRCh37          1   \n",
       "4       OR2T3          343173          genome.wustl.edu     GRCh37          1   \n",
       "\n",
       "   Start_Position  End_Position Strand  \\\n",
       "0        71512366      71512366      +   \n",
       "1       152285981     152285981      +   \n",
       "2       174417411     174417411      +   \n",
       "3       234452419     234452419      +   \n",
       "4       248636826     248636826      +   \n",
       "\n",
       "                              Consequence Variant_Classification  ...  \\\n",
       "0  missense_variant,splice_region_variant      Missense_Mutation  ...   \n",
       "1                        missense_variant      Missense_Mutation  ...   \n",
       "2                      synonymous_variant                 Silent  ...   \n",
       "3                      synonymous_variant                 Silent  ...   \n",
       "4                        missense_variant      Missense_Mutation  ...   \n",
       "\n",
       "  METASTATIC_TUMOR_INDICATOR PROJECT_CODE TISSUE_SOURCE_SITE  \\\n",
       "0                        NaN          NaN                 B6   \n",
       "1                        NaN          NaN                 B6   \n",
       "2                        NaN          NaN                 B6   \n",
       "3                        NaN          NaN                 B6   \n",
       "4                        NaN          NaN                 B6   \n",
       "\n",
       "  TUMOR_TISSUE_SITE                CANCER_TYPE  \\\n",
       "0            Breast  Invasive Breast Carcinoma   \n",
       "1            Breast  Invasive Breast Carcinoma   \n",
       "2            Breast  Invasive Breast Carcinoma   \n",
       "3            Breast  Invasive Breast Carcinoma   \n",
       "4            Breast  Invasive Breast Carcinoma   \n",
       "\n",
       "               CANCER_TYPE_DETAILED ONCOTREE_CODE SAMPLE_TYPE  SOMATIC_STATUS  \\\n",
       "0  Breast Invasive Ductal Carcinoma           IDC     Primary         Matched   \n",
       "1  Breast Invasive Ductal Carcinoma           IDC     Primary         Matched   \n",
       "2  Breast Invasive Ductal Carcinoma           IDC     Primary         Matched   \n",
       "3  Breast Invasive Ductal Carcinoma           IDC     Primary         Matched   \n",
       "4  Breast Invasive Ductal Carcinoma           IDC     Primary         Matched   \n",
       "\n",
       "   TMB_NONSYNONYMOUS  \n",
       "0           1.366667  \n",
       "1           1.366667  \n",
       "2           1.366667  \n",
       "3           1.366667  \n",
       "4           1.366667  \n",
       "\n",
       "[5 rows x 142 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge mutation and clinical data on PATIENT_ID\n",
    "df_merged = pd.merge(df_mutations, df_clinicalsample, on='PATIENT_ID', how='left')\n",
    "df_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cd10ec85-e2b5-4e56-8e64-33439eb34d6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Gene</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTGER3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000356595</td>\n",
       "      <td>ENST00000356595</td>\n",
       "      <td>ENSG00000050628</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLG</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000368799</td>\n",
       "      <td>ENST00000368799</td>\n",
       "      <td>ENSG00000143631</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GPR52</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000367685</td>\n",
       "      <td>ENST00000367685</td>\n",
       "      <td>ENSG00000203737</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLC35F3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000366618</td>\n",
       "      <td>ENST00000366618</td>\n",
       "      <td>ENSG00000183780</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR2T3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000359594</td>\n",
       "      <td>ENST00000359594</td>\n",
       "      <td>ENSG00000196539</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol    PATIENT_ID Variant_Classification Variant_Type  \\\n",
       "0      PTGER3  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "1         FLG  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "2       GPR52  TCGA-B6-A0IG                 Silent          SNP   \n",
       "3     SLC35F3  TCGA-B6-A0IG                 Silent          SNP   \n",
       "4       OR2T3  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "\n",
       "     Transcript_ID          Feature             Gene  \\\n",
       "0  ENST00000356595  ENST00000356595  ENSG00000050628   \n",
       "1  ENST00000368799  ENST00000368799  ENSG00000143631   \n",
       "2  ENST00000367685  ENST00000367685  ENSG00000203737   \n",
       "3  ENST00000366618  ENST00000366618  ENSG00000183780   \n",
       "4  ENST00000359594  ENST00000359594  ENSG00000196539   \n",
       "\n",
       "                 CANCER_TYPE              CANCER_TYPE_DETAILED  \n",
       "0  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "1  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "2  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "3  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "4  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep only the relevant columns\n",
    "cancer = df_merged[['Hugo_Symbol','PATIENT_ID','Variant_Classification',\n",
    "                    'Variant_Type','Transcript_ID','Feature','Gene',\n",
    "                    'CANCER_TYPE','CANCER_TYPE_DETAILED']]\n",
    "cancer.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89a565dd-bfa0-4b20-b2ef-74fdbfa27a8f",
   "metadata": {},
   "source": [
    "##### Here, we selected only the relevant columns from the merged mutation and clinical dataset. These include the gene name (Hugo_Symbol), mutation type (Variant_Classification, Variant_Type), transcript and gene IDs, and cancer subtype information (CANCER_TYPE_DETAILED). This filtered data keeps only the necessary details needed for our classification task and simplifies further processing."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "176bf5d3-15ee-4020-899c-8269d37a2f2d",
   "metadata": {},
   "source": [
    "## PREPROCESSING\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b3e46da-a01b-4ec5-88d1-f163aa905be3",
   "metadata": {},
   "source": [
    "#### Filtering Cancer types:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "32fb3d27-2a14-44ce-b6a7-a7164b56e20e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Filtered shape: (56933, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Gene</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTGER3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000356595</td>\n",
       "      <td>ENST00000356595</td>\n",
       "      <td>ENSG00000050628</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLG</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000368799</td>\n",
       "      <td>ENST00000368799</td>\n",
       "      <td>ENSG00000143631</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GPR52</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000367685</td>\n",
       "      <td>ENST00000367685</td>\n",
       "      <td>ENSG00000203737</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLC35F3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000366618</td>\n",
       "      <td>ENST00000366618</td>\n",
       "      <td>ENSG00000183780</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR2T3</td>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENST00000359594</td>\n",
       "      <td>ENST00000359594</td>\n",
       "      <td>ENSG00000196539</td>\n",
       "      <td>Invasive Breast Carcinoma</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol    PATIENT_ID Variant_Classification Variant_Type  \\\n",
       "0      PTGER3  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "1         FLG  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "2       GPR52  TCGA-B6-A0IG                 Silent          SNP   \n",
       "3     SLC35F3  TCGA-B6-A0IG                 Silent          SNP   \n",
       "4       OR2T3  TCGA-B6-A0IG      Missense_Mutation          SNP   \n",
       "\n",
       "     Transcript_ID          Feature             Gene  \\\n",
       "0  ENST00000356595  ENST00000356595  ENSG00000050628   \n",
       "1  ENST00000368799  ENST00000368799  ENSG00000143631   \n",
       "2  ENST00000367685  ENST00000367685  ENSG00000203737   \n",
       "3  ENST00000366618  ENST00000366618  ENSG00000183780   \n",
       "4  ENST00000359594  ENST00000359594  ENSG00000196539   \n",
       "\n",
       "                 CANCER_TYPE              CANCER_TYPE_DETAILED  \n",
       "0  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "1  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "2  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "3  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  \n",
       "4  Invasive Breast Carcinoma  Breast Invasive Ductal Carcinoma  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep only these two types\n",
    "cancer = cancer[\n",
    "    (cancer['CANCER_TYPE_DETAILED'] == 'Breast Invasive Ductal Carcinoma') |\n",
    "    (cancer['CANCER_TYPE_DETAILED'] == 'Breast Invasive Lobular Carcinoma')\n",
    "]\n",
    "print(\" Filtered shape:\", cancer.shape)\n",
    "cancer.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a786a74a-36eb-4099-8562-c484e1fdce59",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a label mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bfff4648-edea-4428-a0eb-c58c52631e38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-B6-A0IG</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>TCGA-BH-A0HQ</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>TCGA-BH-A18G</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>TCGA-A1-A0SD</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1891</th>\n",
       "      <td>TCGA-A1-A0SF</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        PATIENT_ID              CANCER_TYPE_DETAILED\n",
       "0     TCGA-B6-A0IG  Breast Invasive Ductal Carcinoma\n",
       "45    TCGA-BH-A0HQ  Breast Invasive Ductal Carcinoma\n",
       "84    TCGA-BH-A18G  Breast Invasive Ductal Carcinoma\n",
       "1857  TCGA-A1-A0SD  Breast Invasive Ductal Carcinoma\n",
       "1891  TCGA-A1-A0SF  Breast Invasive Ductal Carcinoma"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Unique patient → label mapping\n",
    "df_target = cancer[['PATIENT_ID', 'CANCER_TYPE_DETAILED']].drop_duplicates()\n",
    "df_target.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12a4c9be-b6d1-4906-b765-1076da9992b8",
   "metadata": {},
   "source": [
    "### One hot-encode Hugo_symbol ( muatated gene column )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c868860c-9922-49fd-b78a-26bce6d18c59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>AADACL4</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15302 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   A2M  A2ML1  A4GALT  A4GNT  AAAS  AACS  AADAC  AADACL2  AADACL4  AADAT  ...  \\\n",
       "0    0      0       0      0     0     0      0        0        0      0  ...   \n",
       "1    0      0       0      0     0     0      0        0        0      0  ...   \n",
       "2    0      0       0      0     0     0      0        0        0      0  ...   \n",
       "3    0      0       0      0     0     0      0        0        0      0  ...   \n",
       "4    0      0       0      0     0     0      0        0        0      0  ...   \n",
       "\n",
       "   ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  ZYX  ZZEF1  ZZZ3  \\\n",
       "0      0     0     0     0       0       0    0      0     0   \n",
       "1      0     0     0     0       0       0    0      0     0   \n",
       "2      0     0     0     0       0       0    0      0     0   \n",
       "3      0     0     0     0       0       0    0      0     0   \n",
       "4      0     0     0     0       0       0    0      0     0   \n",
       "\n",
       "               CANCER_TYPE_DETAILED  \n",
       "0  Breast Invasive Ductal Carcinoma  \n",
       "1  Breast Invasive Ductal Carcinoma  \n",
       "2  Breast Invasive Ductal Carcinoma  \n",
       "3  Breast Invasive Ductal Carcinoma  \n",
       "4  Breast Invasive Ductal Carcinoma  \n",
       "\n",
       "[5 rows x 15302 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Turn each gene into a binary column\n",
    "dummies = pd.get_dummies(cancer[\"Hugo_Symbol\"], drop_first=True)\n",
    "\n",
    "# Convert boolean True/False to 1/0\n",
    "dummies = dummies.astype(int)\n",
    "\n",
    "# Add cancer type back\n",
    "dummies = pd.concat([dummies, cancer[\"CANCER_TYPE_DETAILED\"]], axis=1)\n",
    "dummies.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23247e81-33d7-47f9-9164-be7bdf19f100",
   "metadata": {},
   "source": [
    "##### we transformed the Hugo_Symbol column (which contains gene names) into binary features using one-hot encoding. Each gene is turned into a separate column with a value of 1 if it is mutated in that row, and 0 otherwise. We then added the CANCER_TYPE_DETAILED column back to associate each mutation profile with its corresponding cancer subtype. This prepares the dataset for classification by converting categorical gene mutation data into a numeric format that machine learning models can understand."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "723a88b3-9b80-4bd8-b462-d1a85bfd915b",
   "metadata": {},
   "source": [
    "### Combine Patien Id with Mutaion and group "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "32bd7361-fb85-4e27-8889-932101410df0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>AADACL4</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SD</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SE</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SI</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SJ</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15302 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              A2M  A2ML1  A4GALT  A4GNT  AAAS  AACS  AADAC  AADACL2  AADACL4  \\\n",
       "PATIENT_ID                                                                     \n",
       "TCGA-A1-A0SD    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SE    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SF    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SI    0      0       0      0     0     0      0        0        0   \n",
       "TCGA-A1-A0SJ    0      0       0      0     0     0      0        0        0   \n",
       "\n",
       "              AADAT  ...  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  ZYX  ZZEF1  \\\n",
       "PATIENT_ID           ...                                                        \n",
       "TCGA-A1-A0SD      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SE      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SF      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SI      0  ...      0     0     0     0       0       0    0      0   \n",
       "TCGA-A1-A0SJ      0  ...      0     0     0     0       0       0    0      0   \n",
       "\n",
       "              ZZZ3               CANCER_TYPE_DETAILED  \n",
       "PATIENT_ID                                             \n",
       "TCGA-A1-A0SD     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-A1-A0SE     0  Breast Invasive Lobular Carcinoma  \n",
       "TCGA-A1-A0SF     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-A1-A0SI     0   Breast Invasive Ductal Carcinoma  \n",
       "TCGA-A1-A0SJ     0   Breast Invasive Ductal Carcinoma  \n",
       "\n",
       "[5 rows x 15302 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Combine with patient ID\n",
    "df = pd.concat([cancer[['PATIENT_ID']], dummies], axis=1)\n",
    "\n",
    "# Group by patient and take max (1 if gene mutated at least once)\n",
    "sparse_df = df.groupby(\"PATIENT_ID\")[dummies.columns].max()\n",
    "\n",
    "# Just in case: remove any duplicated columns\n",
    "sparse_df = sparse_df.loc[:, ~sparse_df.columns.duplicated()]\n",
    "\n",
    "display(sparse_df.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8475f12-400c-4ac7-ae52-cfddad45f773",
   "metadata": {},
   "source": [
    "### Reset index and prprare X andY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "604ceda4-fc48-4ba0-9a62-6a49808160ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " X shape: (617, 15301)\n",
      " y shape: (617,)\n"
     ]
    }
   ],
   "source": [
    "# Reset index so PATIENT_ID is a column again\n",
    "df_logir = sparse_df.copy()\n",
    "df_logir.reset_index(inplace=True)\n",
    "\n",
    "# Define X (features)\n",
    "x = df_logir.drop(['CANCER_TYPE_DETAILED', 'PATIENT_ID'], axis=1)\n",
    "\n",
    "# Define y (labels)\n",
    "y = df_logir['CANCER_TYPE_DETAILED']\n",
    "\n",
    "print(\" X shape:\", x.shape)\n",
    "print(\" y shape:\", y.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4e675465-580d-4db0-91d5-713111b62ec4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CANCER_TYPE_DETAILED\n",
       "Breast Invasive Ductal Carcinoma     490\n",
       "Breast Invasive Lobular Carcinoma    127\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y.value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ada39d2-d945-49e1-8173-2d664ab1569f",
   "metadata": {},
   "source": [
    "####  we reset the index to bring PATIENT_ID back as a regular column and then prepared the input (X) and target (y) for our classification task.\n",
    "\n",
    "X contains binary features representing the presence or absence of mutations in 15,301 genes for each patient.\n",
    "\n",
    "y contains the corresponding cancer subtype label for each patient — either ductal or lobular.\n",
    "\n",
    "We have a total of 617 patients, with 490 cases of ductal carcinoma and 127 cases of lobular carcinoma, making this a binary classification problem with imbalanced classes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71907033-f802-4b16-b7b2-0fdb06b16682",
   "metadata": {},
   "source": [
    "### CrossValidation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9cab201b-2e07-4677-b6ed-14197518c83c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " x_scaled shape: (617, 15301)\n"
     ]
    }
   ],
   "source": [
    "scaler = StandardScaler()\n",
    "x_scaled = scaler.fit_transform(x)\n",
    "\n",
    "print(\" x_scaled shape:\", x_scaled.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3991453b-7d96-444d-83d1-46119297ec34",
   "metadata": {},
   "source": [
    "#### Set  Evaluation metrics and 5 fold CV\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "594728d4-7a0a-44e9-99c6-c64e39e61c5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scoring metrics\n",
    "scoring = {\n",
    "    'accuracy': make_scorer(accuracy_score),\n",
    "    'precision': make_scorer(precision_score),\n",
    "    'recall': make_scorer(recall_score),\n",
    "    'f1_score': make_scorer(f1_score)\n",
    "}\n",
    "\n",
    "# 5-fold cross-validation\n",
    "skf = KFold(n_splits=5, shuffle=True, random_state=42)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "395b2dd8-4b29-4ef9-98ca-5bd49e73102a",
   "metadata": {},
   "source": [
    "### logistic regressiion\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0e6f3575-766a-43ea-81d9-18d40cfaac1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "logir_ovr = LogisticRegression(random_state=0, multi_class='ovr', max_iter=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "32c33319-0c05-4afb-8a55-3cf89bd1d68d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Evaluation metrics\n",
    "logit_accuracies, logit_precision, logit_recall, logit_f1scores = [], [], [], []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b6896c3-c405-49ec-8984-22c0ce9c8046",
   "metadata": {},
   "source": [
    "### Perform cross val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "27692a73-1a13-459d-84e3-50eb005e4cbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logistic Regression F1 Scores: [0.44144144144144143, 0.4439461883408072, 0.4434389140271493, 0.4533333333333333, 0.4305555555555556]\n"
     ]
    }
   ],
   "source": [
    "# Perform cross-validation\n",
    "for train_idx, test_idx in skf.split(x_scaled, y):\n",
    "    x_train, x_test = x_scaled[train_idx], x_scaled[test_idx]\n",
    "    y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "    \n",
    "    logir_ovr.fit(x_train, y_train)\n",
    "    y_pred = logir_ovr.predict(x_test)\n",
    "    \n",
    "    logit_accuracies.append(logir_ovr.score(x_test, y_test))\n",
    "    logit_f1scores.append(metrics.f1_score(y_test, y_pred, average='macro'))\n",
    "    logit_precision.append(metrics.precision_score(y_test, y_pred, average='macro'))\n",
    "    logit_recall.append(metrics.recall_score(y_test, y_pred, average='macro'))\n",
    "\n",
    "# Show results\n",
    "print(\"Logistic Regression F1 Scores:\", logit_f1scores)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22912572-bf05-4d3f-a64e-d89a5acc9056",
   "metadata": {},
   "source": [
    "### In this step, we performed 5-fold cross-validation using a logistic regression model to evaluate how well it can classify the two breast cancer subtypes based on mutation profiles.\n",
    "\n",
    "For each fold:\n",
    "\n",
    "The data was split into training and testing sets.\n",
    "\n",
    "The model was trained on the training set and tested on the test set.\n",
    "\n",
    "We recorded key evaluation metrics: accuracy, precision, recall, and F1 score (macro-averaged).\n",
    "\n",
    "The F1 scores across the 5 folds show consistent performance, with scores ranging around 0.43–0.45, indicating that the model is moderately effective but likely impacted by class imbalance and high dimensionality."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5647e103-1da5-4d8b-ab56-c1852dc72191",
   "metadata": {},
   "source": [
    "## Naive Bayes classifcation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cf07758a-2f49-4479-9add-29496cc3041e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model\n",
    "naive_bayes = BernoulliNB()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "00315fcc-b170-41ce-ba19-03e82ffc41bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Evaluation metric lists\n",
    "naive_accuracies, naive_precision, naive_recall, naive_f1scores = [], [], [], []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2bfff130-223d-436a-b942-37d889d4dee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5-fold CV\n",
    "for train_idx, test_idx in skf.split(x_scaled, y):\n",
    "    x_train, x_test = x_scaled[train_idx], x_scaled[test_idx]\n",
    "    y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "    \n",
    "    naive_bayes.fit(x_train, y_train)\n",
    "    y_pred = naive_bayes.predict(x_test)\n",
    "    \n",
    "    naive_accuracies.append(naive_bayes.score(x_test, y_test))\n",
    "    naive_f1scores.append(metrics.f1_score(y_test, y_pred, average='macro'))\n",
    "    naive_precision.append(metrics.precision_score(y_test, y_pred, average='macro'))\n",
    "    naive_recall.append(metrics.recall_score(y_test, y_pred, average='macro'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9ec13c21-7b56-49bf-a216-fdcb7c9b8fc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Naive Bayes F1 Scores: [0.4976851851851852, 0.5070336391437309, 0.48806881243063266, 0.4722707770538694, 0.45394006659267483]\n"
     ]
    }
   ],
   "source": [
    "print(\"Naive Bayes F1 Scores:\", naive_f1scores)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fce7971-c130-42eb-87bf-d6170a53e6bd",
   "metadata": {},
   "source": [
    "#### we trained a Bernoulli Naive Bayes classifier using 5-fold cross-validation. This model is well-suited for binary input data, like our mutation presence/absence matrix.\n",
    "\n",
    "For each fold:\n",
    "\n",
    "The model was trained on the training split and predictions were made on the test split.\n",
    "\n",
    "We calculated accuracy, precision, recall, and macro-averaged F1 score.\n",
    "\n",
    "The F1 scores ranged from ~0.45 to 0.51, which are higher than logistic regression, indicating that Naive Bayes handled the high-dimensional sparse mutation data better in this case."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6e04abe-9f3c-4cbb-8510-f0769b01338e",
   "metadata": {},
   "source": [
    "## LDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c256aa97-f402-46f9-9689-69e076eb43dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LDA F1 Scores: [0.44144144144144143, 0.3721312553429342, 0.41380718954248363, 0.47435897435897434, 0.4305555555555556]\n"
     ]
    }
   ],
   "source": [
    "# Initialize LDA model\n",
    "lda = LinearDiscriminantAnalysis()\n",
    "\n",
    "# Metric lists\n",
    "lda_accuracies, lda_precision, lda_recall, lda_f1scores = [], [], [], []\n",
    "\n",
    "# Cross-validation loop\n",
    "for train_idx, test_idx in skf.split(x_scaled, y):\n",
    "    x_train, x_test = x_scaled[train_idx], x_scaled[test_idx]\n",
    "    y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]\n",
    "    \n",
    "    lda.fit(x_train, y_train)\n",
    "    y_pred = lda.predict(x_test)\n",
    "    \n",
    "    lda_accuracies.append(lda.score(x_test, y_test))\n",
    "    lda_f1scores.append(metrics.f1_score(y_test, y_pred, average='macro'))\n",
    "    lda_precision.append(metrics.precision_score(y_test, y_pred, average='macro'))\n",
    "    lda_recall.append(metrics.recall_score(y_test, y_pred, average='macro'))\n",
    "\n",
    "# Show results\n",
    "print(\"LDA F1 Scores:\", lda_f1scores)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1bb02c8-b5ca-4ff4-97ef-d456ef1d79b3",
   "metadata": {},
   "source": [
    "###  We applied Linear Discriminant Analysis (LDA) to classify the breast cancer subtypes using 5-fold cross-validation.\n",
    "\n",
    "LDA attempts to find a linear combination of features that best separates the two classes. However, it may not perform optimally on high-dimensional sparse data like gene mutations.\n",
    "\n",
    "The F1 scores varied across folds, ranging from ~0.37 to 0.47.\n",
    "\n",
    "This performance was generally lower and more inconsistent compared to Naive Bayes, suggesting LDA struggled with the feature space or class imbalance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "07d7e247-8e9d-45b6-905f-0e6ab6735df3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Average F1 Scores:\n",
      "Logistic Regression: 0.4425\n",
      "Naive Bayes: 0.4838\n",
      "LDA: 0.4265\n",
      "\n",
      " Best performing model based on F1 score:\n",
      " Naive Bayes with F1 score: 0.4838\n"
     ]
    }
   ],
   "source": [
    "def average(scores):\n",
    "    return sum(scores) / len(scores)\n",
    "\n",
    "# Store model names and their average F1 scores\n",
    "model_names = ['Logistic Regression', 'Naive Bayes', 'LDA']\n",
    "avg_f1s = [\n",
    "    average(logit_f1scores),\n",
    "    average(naive_f1scores),\n",
    "    average(lda_f1scores)\n",
    "]\n",
    "\n",
    "# Print averages\n",
    "print(\" Average F1 Scores:\")\n",
    "for name, score in zip(model_names, avg_f1s):\n",
    "    print(f\"{name}: {round(score, 4)}\")\n",
    "\n",
    "# Find best model\n",
    "best_index = avg_f1s.index(max(avg_f1s))\n",
    "print(\"\\n Best performing model based on F1 score:\")\n",
    "print(f\" {model_names[best_index]} with F1 score: {round(avg_f1s[best_index], 4)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af95023b-cc02-4c8b-b302-9b30aa0ecbbc",
   "metadata": {},
   "source": [
    "### Comparision\n",
    "### Based on the average F1 scores from 5-fold cross-validation, Naive Bayes achieved the best performance with a score of 0.4838, followed by Logistic Regression at 0.4425, and LDA at 0.4265.\n",
    "\n",
    "### This suggests that Naive Bayes was more effective in handling our high-dimensional, sparse binary data (gene mutation presence/absence), likely due to its assumption of feature independence. Logistic Regression also performed reasonably well, while LDA showed the weakest performance, likely because it is less suited for sparse data with a large number of features compared to samples."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
